### ISSUE TITLE
[FEATURE] Timestamp from string

### ISSUE THREAD

**Issue by MaximilianSchon** [dynamoose/dynamoose#1092]
### Summary:
When sending a timestamp as JSON to and from a JSON API it gets converted to a string, which is reasonable as JSON does not inherently support dates. This means when getting the date back from a client, it is a string, which results in dynamoose rejecting the timestamp if the entire object is passed into the update function. 

In this scenario I think it makes more sense for dynamoose to do a check if the field is a timestamp-field and convert (attempt to) convert it to a valid date, instead of rejecting it. 

Possibly #977 solves this issue. 

### Other:
- [x] I have read through the Dynamoose documentation before posting this issue
- [x] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this feature has not already been suggested before
- [x] I have filled out all fields above
- [x] I am running the latest version of Dynamoose

**Comment by fishcharlie** [comment]
@MaximilianSchon This is a tough one. The big question here is how far should Dynamoose interfere with your data? Converting a string to a date seems extremely excessive for this library in my opinion. There are so many different edge cases and different patterns that could exist here. I think Node/JS does support taking in a string and outputting a date object. But again, it seems excessive.

I'm not normally a fan of black box libraries. Meaning, doing behind the scenes stuff that isn't obvious or clear at all. For the most part in this library I try to avoid that as much as possible. Adding logic that isn't obvious or clear makes things a lot harder to debug for users of Dynamoose. What happens if it automatically gets converted to a date but isn't the right date? Or a string gets passed in that looks like a date and gets parsed to a date but the user wants to store as a string?

You also mention #977. Which isn't related to this issue at all. **However** that is the ideal solution here. Changing how `set` works, and allowing you to pass in your own function to convert from String to Date. I think we might also need to make changes to allow this for timestamps tho. Currently I don't think it's possible to do that for timestamps.

In the meantime you can always just change your data before passing it into Dynamoose. So it requires more code on your end, but it shouldn't block anything you are doing.

Hopefully this makes sense with my thought process here. If you have counter arguments or other reasons for why you think this behavior should be changed/implemented, please comment here. Having back and forth discussion and talking about the pros and cons is important in order to make the best decisions.

**Comment by MaximilianSchon** [comment]
> @MaximilianSchon This is a tough one. The big question here is how far should Dynamoose interfere with your data? Converting a string to a date seems extremely excessive for this library in my opinion. There are so many different edge cases and different patterns that could exist here. I think Node/JS does support taking in a string and outputting a date object. But again, it seems excessive.
> 
> I'm not normally a fan of black box libraries. Meaning, doing behind the scenes stuff that isn't obvious or clear at all. For the most part in this library I try to avoid that as much as possible. Adding logic that isn't obvious or clear makes things a lot harder to debug for users of Dynamoose. What happens if it automatically gets converted to a date but isn't the right date? Or a string gets passed in that looks like a date and gets parsed to a date but the user wants to store as a string?
> 
> You also mention #977. Which isn't related to this issue at all. **However** that is the ideal solution here. Changing how `set` works, and allowing you to pass in your own function to convert from String to Date. I think we might also need to make changes to allow this for timestamps tho. Currently I don't think it's possible to do that for timestamps.
> 
> In the meantime you can always just change your data before passing it into Dynamoose. So it requires more code on your end, but it shouldn't block anything you are doing.
> 
> Hopefully this makes sense with my thought process here. If you have counter arguments or other reasons for why you think this behavior should be changed/implemented, please comment here. Having back and forth discussion and talking about the pros and cons is important in order to make the best decisions.

So the behaviour I am after is simply updating the timestamp without having to delete the fields myself, as it would minimize the code I have to write. 

The problem if dynamoose deletes the field, is that it might end up with someone passing a property to dynamoose that they want to save, and not detecting it, since they are not getting an exception. 

Perhaps a function like this one: https://validatejs.org/#utilities-clean-attributes, that is callable on the Schema object? This would make it opt-in, it would make sure that all your saves work and the clean function could return a list of which attributes where filtered out so that they could be observed? It would also make for more specific error messages and error handling. The data that where filtered out would instead obtain their default value.

I do not know if this is the ambition with dynamoose though, and the other scenario is that I simply use a validation library, but since dynamoose already does validation I feel like that there is an ambition to make sure that the data is correct.

**Comment by fishcharlie** [comment]
@MaximilianSchon 

> So the behaviour I am after is simply updating the timestamp without having to delete the fields myself
> The problem if dynamoose deletes the field

Don't understand what you mean here. Your original post doesn't mention deleting at all, and Dynamoose doesn't delete fields unless they aren't in your schema.

> is that it might end up with someone passing a property to dynamoose that they want to save, and not detecting it, since they are not getting an exception

I'm curious more about your use case here. Are you using `saveUnknown`? Dynamoose is really meant to add structure to your data. Users really shouldn't be passing in properties that you aren't expecting. Now if it's a typo or mistake, that is something that Dynamoose could improve on. However that is a completely different topic than your original post. So not sure I'm following the purpose of this.

Ideally there might be a strict option that would throw an error if the property is not in your schema and you are trying to add it? That might be a cool feature. But again, seems completely off topic from the original post.

**Comment by MaximilianSchon** [comment]
> @MaximilianSchon
> 
> > So the behaviour I am after is simply updating the timestamp without having to delete the fields myself
> > The problem if dynamoose deletes the field
> 
> Don't understand what you mean here. Your original post doesn't mention deleting at all, and Dynamoose doesn't delete fields unless they aren't in your schema.
> 

I am sorry for not being very coherent. It's pretty late and I have had 2 seperate but related issues. 

I recently noticed that batchPut was not updating my timestamps, which I think is correct behavior, so I had to update how I did my saving, to use the update function. Not sure why batchPut did not complain 
about types though.  I can look more into this as well tomorrow. 

Changing lead me to the following conclusions:
- Type mismatch, I am trying to put a string into a date. (This is when I created this issue) I don't really want to save the date, I just want to get rid of the exception, as the date seems to be set afterwards anyway, but I should likely not rely on implementation details for this, so this was not a very good solution from me anyway. 
- forcing updatedAt to a date, threw an exception which I think indicated that updatedAt was sent to dynamodb twice. I can give you example code tomorrow.  In this case I gave it new Date(). 
- deleting the variable updatedAt resulted in the correct behavior

> > is that it might end up with someone passing a property to dynamoose that they want to save, and not detecting it, since they are not getting an exception
> 
> I'm curious more about your use case here. Are you using `saveUnknown`? Dynamoose is really meant to add structure to your data. Users really shouldn't be passing in properties that you aren't expecting. Now if it's a typo or mistake, that is something that Dynamoose could improve on. However that is a completely different topic than your original post. So not sure I'm following the purpose of this.
> 

I was just thinking generally for dynamoose and not for my specific use case. For my use case removing the fields would likely be optimal. I was just trying to look for cons with that approach and error detection would be one reason against deleting fields that do not follow the schema. 
I am not using saveUnknown.  

> Ideally there might be a strict option that would throw an error if the property is not in your schema and you are trying to add it? That might be a cool feature. But again, seems completely off topic from the original post.

I am probably leaning closer to an option where non-strict would save the correct fields and not throw any exceptions about the incorrect fields. But perhaps this is what you were thinking as well? 

With this approach I can send updatedAt from my frontend as a string, dynamoose will filter it out and the field will update as it should. 

To clarify: 
I fetch the object from my backend, dynamoose gets it and sends it as json to my frontend. 
My frontend allows me to update the fields and sends back the entire object as it came, with modified values.

But when sending it back and forth the data conversion makes the data incorrect and the schema validation does not pass. 

I think perhaps the dynamoose update is more related to PATCH and I built my API based on how one would using PUT, if that comparison make sense to you? 

----

What about having an attribute in the schema that states an action if the validation fails? The options would likely be to use the default or throw an exception. I do realise however that this requires that the validation runs twice which feels unnecessary.
